### PR TITLE
Thin `AssocItem`

### PR DESCRIPTION
**Pull request by cjgillot**
This PR removes a few fields from `AssocItem` that should be easily computed using other queries.
This simplifies some of the metadata decoding.

### CODE REVIEW

**Comment by cjgillot** [comment]
r? rust-lang/compiler

**Review by lcnr — APPROVED** [review]

File: src/test/incremental/hashes/inherent_impls.rs [hunk 1]
```diff
@@ -116,9 +116,9 @@ impl Foo {
 // Change Method Privacy -------------------------------------------------------
 #[cfg(any(cfail1,cfail4))]
 impl Foo {
-    //----------------------------------------------------
     //--------------------------
-    //------------------------------------------------------------------------------
+    //--------------------------
+    //--------------------------------------------------------------
     //--------------------------
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply src/test/incremental/hashes/inherent_impls.rs:122]: The point is to have all the versions of the impl block be long of the same number of bytes. This allows to reduce the amount of span-based invalidations in `incremental_relative_spans` mode, and test "optimal" incremental reuse.

**Comment by nnethercote** [comment]
This looks perf-neutral, with the improvements roughly the same size as the regressions.

@rustbot label: +perf-regression-triaged

**Comment by ethe** [comment]
get panic in any cargo projects building after this PR.
panic logs:
```
 ↳ cargo build
   Compiling research v0.1.0 (/data00/home/autohime/research)
error: internal compiler error: encountered incremental compilation error with late_bound_vars_map(research[9e26]::_call)
  |
  = help: This is a known issue with the compiler. Run `cargo clean -p research` or `cargo clean` to allow your project to compile
  = note: Please follow the instructions below to create a bug report with the provided information
  = note: See <https://github.com/rust-lang/rust/issues/84970> for more information

thread 'rustc' panicked at 'Found unstable fingerprints for late_bound_vars_map(research[9e26]::_call): Some({0: [], 61: [], 20: [], 81: [], 72: [Region(BrNamed(DefId(0:9 ~ research[9e26]::_call::'s), 's)), Region(BrNamed(DefId(0:10 ~ research[9e26]::_call::'cx), 'cx))], 69: [], 66: [], 79: [Region(BrNamed(DefId(0:9 ~ research[9e26]::_call::'s), 's)), Region(BrNamed(DefId(0:10 ~ research[9e26]::_call::'cx), 'cx))]})', compiler/rustc_query_system/src/query/plumbing.rs:627:9
stack backtrace:
   0:     0x7fb4bd05e474 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1525352d64489be5
   1:     0x7fb4bd0d788e - core::fmt::write::he0cdbcb1b4f9137e
   2:     0x7fb4bd0475a1 - std::io::Write::write_fmt::h5ac120bcc89a9711
   3:     0x7fb4bd05e2c9 - std::sys_common::backtrace::print::h357844ba9e23590a
   4:     0x7fb4bd0480a7 - std::panicking::default_hook::{{closure}}::h046e90d133347414
   5:     0x7fb4bd047ed3 - std::panicking::default_hook::h5e50f9108c63488b
   6:     0x7fb4bd8f4f74 - rustc_driver[a177c53ac2cf0051]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:     0x7fb4bd0484a8 - std::panicking::rust_panic_with_hook::hb9114b8583c4cdf6
   8:     0x7fb4bd0564e7 - std::panicking::begin_panic_handler::{{closure}}::h7518f7cb4822dfdc
   9:     0x7fb4bd056424 - std::sys_common::backtrace::__rust_end_short_backtrace::he6f9aab0edfe27da
  10:     0x7fb4bd048172 - rust_begin_unwind
  11:     0x7fb4bd02a0d3 - core::panicking::panic_fmt::hb8c3375d42574ff4
  12:     0x7fb4bd8cc300 - rustc_query_system[5961241f39e0d58f]::query::plumbing::incremental_verify_ich_cold
  13:     0x7fb4bed71569 - rustc_query_system[5961241f39e0d58f]::query::plumbing::incremental_verify_ich::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt, rustc_span[18a919f8c0850d33]::def_id::LocalDefId, core[5571ea7c97e175b]::option::Option<&std[a2c7089b4b8c7571]::collections::hash::map::HashMap<rustc_hir[bd57e2855bf4af64]::hir_id::ItemLocalId, alloc[d30d5095469b2716]::vec::Vec<rustc_middle[5d1ff1ffbdb84c71]::ty::sty::BoundVariableKind>, core[5571ea7c97e175b]::hash::BuildHasherDefault<rustc_hash[4aa44f1efe9a28e9]::FxHasher>>>>
  14:     0x7fb4bedaa4f8 - rustc_query_system[5961241f39e0d58f]::query::plumbing::try_load_from_disk_and_cache_in_memory::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt, rustc_span[18a919f8c0850d33]::def_id::LocalDefId, core[5571ea7c97e175b]::option::Option<&std[a2c7089b4b8c7571]::collections::hash::map::HashMap<rustc_hir[bd57e2855bf4af64]::hir_id::ItemLocalId, alloc[d30d5095469b2716]::vec::Vec<rustc_middle[5d1ff1ffbdb84c71]::ty::sty::BoundVariableKind>, core[5571ea7c97e175b]::hash::BuildHasherDefault<rustc_hash[4aa44f1efe9a28e9]::FxHasher>>>>
  15:     0x7fb4bed18bad - rustc_query_system[5961241f39e0d58f]::query::plumbing::try_execute_query::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt, rustc_query_system[5961241f39e0d58f]::query::caches::DefaultCache<rustc_span[18a919f8c0850d33]::def_id::LocalDefId, core[5571ea7c97e175b]::option::Option<&std[a2c7089b4b8c7571]::collections::hash::map::HashMap<rustc_hir[bd57e2855bf4af64]::hir_id::ItemLocalId, alloc[d30d5095469b2716]::vec::Vec<rustc_middle[5d1ff1ffbdb84c71]::ty::sty::BoundVariableKind>, core[5571ea7c97e175b]::hash::BuildHasherDefault<rustc_hash[4aa44f1efe9a28e9]::FxHasher>>>>>
  16:     0x7fb4bede9ba4 - rustc_query_system[5961241f39e0d58f]::query::plumbing::get_query::<rustc_query_impl[b30ae0b18ffa5f8d]::queries::late_bound_vars_map, rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  17:     0x7fb4bec9291d - <rustc_query_impl[b30ae0b18ffa5f8d]::Queries as rustc_middle[5d1ff1ffbdb84c71]::ty::query::QueryEngine>::late_bound_vars_map
  18:     0x7fb4bfabddf8 - <rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt>::late_bound_vars
  19:     0x7fb4be29c798 - rustc_typeck[2b53629f01fad2c1]::collect::gather_explicit_predicates_of
  20:     0x7fb4be29e1d0 - rustc_typeck[2b53629f01fad2c1]::collect::explicit_predicates_of
  21:     0x7fb4bee3ba55 - <rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind as rustc_query_system[5961241f39e0d58f]::dep_graph::DepKind>::with_deps::<<rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::with_task_impl<rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt, rustc_span[18a919f8c0850d33]::def_id::DefId, rustc_middle[5d1ff1ffbdb84c71]::ty::generics::GenericPredicates>::{closure#0}, rustc_middle[5d1ff1ffbdb84c71]::ty::generics::GenericPredicates>
  22:     0x7fb4bef64522 - <rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt, rustc_span[18a919f8c0850d33]::def_id::DefId, rustc_middle[5d1ff1ffbdb84c71]::ty::generics::GenericPredicates>
  23:     0x7fb4bed3fb0b - rustc_query_system[5961241f39e0d58f]::query::plumbing::try_execute_query::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt, rustc_query_system[5961241f39e0d58f]::query::caches::DefaultCache<rustc_span[18a919f8c0850d33]::def_id::DefId, rustc_middle[5d1ff1ffbdb84c71]::ty::generics::GenericPredicates>>
  24:     0x7fb4becdf16a - rustc_query_system[5961241f39e0d58f]::query::plumbing::force_query::<rustc_query_impl[b30ae0b18ffa5f8d]::queries::explicit_predicates_of, rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  25:     0x7fb4bf09d9c6 - rustc_query_impl[b30ae0b18ffa5f8d]::query_callbacks::explicit_predicates_of::force_from_dep_node
  26:     0x7fb4bfaa6d97 - <rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt as rustc_query_system[5961241f39e0d58f]::dep_graph::DepContext>::try_force_from_dep_node
  27:     0x7fb4bef365e6 - <rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  28:     0x7fb4bef36432 - <rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  29:     0x7fb4beef975c - <rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::try_mark_green::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  30:     0x7fb4bece83c8 - rustc_query_system[5961241f39e0d58f]::query::plumbing::ensure_must_run::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt, rustc_span[18a919f8c0850d33]::def_id::DefId, alloc[d30d5095469b2716]::vec::Vec<&rustc_middle[5d1ff1ffbdb84c71]::mir::coverage::CodeRegion>>
  31:     0x7fb4bedd3ea2 - rustc_query_system[5961241f39e0d58f]::query::plumbing::get_query::<rustc_query_impl[b30ae0b18ffa5f8d]::queries::predicates_of, rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  32:     0x7fb4bec6071e - <rustc_query_impl[b30ae0b18ffa5f8d]::Queries as rustc_middle[5d1ff1ffbdb84c71]::ty::query::QueryEngine>::predicates_of
  33:     0x7fb4be2933c1 - rustc_typeck[2b53629f01fad2c1]::collect::convert_item
  34:     0x7fb4be28e98a - <rustc_typeck[2b53629f01fad2c1]::collect::CollectItemTypesVisitor as rustc_hir[bd57e2855bf4af64]::intravisit::Visitor>::visit_item
  35:     0x7fb4be353220 - <rustc_middle[5d1ff1ffbdb84c71]::hir::map::Map>::visit_item_likes_in_module::<rustc_typeck[2b53629f01fad2c1]::collect::CollectItemTypesVisitor>
  36:     0x7fb4be28e0cd - rustc_typeck[2b53629f01fad2c1]::collect::collect_mod_item_types
  37:     0x7fb4bee39c2d - <rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind as rustc_query_system[5961241f39e0d58f]::dep_graph::DepKind>::with_deps::<<rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::with_task_impl<rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt, rustc_span[18a919f8c0850d33]::def_id::LocalDefId, ()>::{closure#0}, ()>
  38:     0x7fb4bef56b5c - <rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt, rustc_span[18a919f8c0850d33]::def_id::LocalDefId, ()>
  39:     0x7fb4bed25a5f - rustc_query_system[5961241f39e0d58f]::query::plumbing::try_execute_query::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt, rustc_query_system[5961241f39e0d58f]::query::caches::DefaultCache<rustc_span[18a919f8c0850d33]::def_id::LocalDefId, ()>>
  40:     0x7fb4bedf41c4 - rustc_query_system[5961241f39e0d58f]::query::plumbing::get_query::<rustc_query_impl[b30ae0b18ffa5f8d]::queries::collect_mod_item_types, rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  41:     0x7fb4bec74f7d - <rustc_query_impl[b30ae0b18ffa5f8d]::Queries as rustc_middle[5d1ff1ffbdb84c71]::ty::query::QueryEngine>::collect_mod_item_types
  42:     0x7fb4be352887 - <rustc_middle[5d1ff1ffbdb84c71]::hir::map::Map>::for_each_module::<rustc_typeck[2b53629f01fad2c1]::check_crate::{closure#0}::{closure#0}::{closure#0}>
  43:     0x7fb4be332772 - <rustc_session[5fd84be164c87cb9]::session::Session>::track_errors::<rustc_typeck[2b53629f01fad2c1]::check_crate::{closure#0}, ()>
  44:     0x7fb4be0ff441 - rustc_typeck[2b53629f01fad2c1]::check_crate
  45:     0x7fb4bda432a1 - rustc_interface[a80cab810099d0cd]::passes::analysis
  46:     0x7fb4bee409f4 - <rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind as rustc_query_system[5961241f39e0d58f]::dep_graph::DepKind>::with_deps::<<rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::with_task_impl<rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>
  47:     0x7fb4bef8597a - <rustc_query_system[5961241f39e0d58f]::dep_graph::graph::DepGraph<rustc_middle[5d1ff1ffbdb84c71]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[5d1ff1ffbdb84c71]::ty::context::TyCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>
  48:     0x7fb4bed5a157 - rustc_query_system[5961241f39e0d58f]::query::plumbing::try_execute_query::<rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt, rustc_query_system[5961241f39e0d58f]::query::caches::DefaultCache<(), core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>>
  49:     0x7fb4bee1ce42 - rustc_query_system[5961241f39e0d58f]::query::plumbing::get_query::<rustc_query_impl[b30ae0b18ffa5f8d]::queries::analysis, rustc_query_impl[b30ae0b18ffa5f8d]::plumbing::QueryCtxt>
  50:     0x7fb4bec5fa72 - <rustc_query_impl[b30ae0b18ffa5f8d]::Queries as rustc_middle[5d1ff1ffbdb84c71]::ty::query::QueryEngine>::analysis
  51:     0x7fb4bd95a724 - <rustc_interface[a80cab810099d0cd]::passes::QueryContext>::enter::<rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>
  52:     0x7fb4bd955e18 - <rustc_interface[a80cab810099d0cd]::interface::Compiler>::enter::<rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}::{closure#2}, core[5571ea7c97e175b]::result::Result<core[5571ea7c97e175b]::option::Option<rustc_interface[a80cab810099d0cd]::queries::Linker>, rustc_errors[c08561276749af2a]::ErrorGuaranteed>>
  53:     0x7fb4bd9666a7 - rustc_span[18a919f8c0850d33]::with_source_map::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, rustc_interface[a80cab810099d0cd]::interface::create_compiler_and_run<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}>::{closure#1}>
  54:     0x7fb4bd956993 - rustc_interface[a80cab810099d0cd]::interface::create_compiler_and_run::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}>
  55:     0x7fb4bd969bf2 - <scoped_tls[cbbb43b5eb84c7ca]::ScopedKey<rustc_span[18a919f8c0850d33]::SessionGlobals>>::set::<rustc_interface[a80cab810099d0cd]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>
  56:     0x7fb4bd9003c9 - std[a2c7089b4b8c7571]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[a80cab810099d0cd]::util::run_in_thread_pool_with_globals<rustc_interface[a80cab810099d0cd]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>
  57:     0x7fb4bd8ff45e - std[a2c7089b4b8c7571]::panicking::try::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, core[5571ea7c97e175b]::panic::unwind_safe::AssertUnwindSafe<<std[a2c7089b4b8c7571]::thread::Builder>::spawn_unchecked_<rustc_interface[a80cab810099d0cd]::util::run_in_thread_pool_with_globals<rustc_interface[a80cab810099d0cd]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  58:     0x7fb4bd953db0 - <<std[a2c7089b4b8c7571]::thread::Builder>::spawn_unchecked_<rustc_interface[a80cab810099d0cd]::util::run_in_thread_pool_with_globals<rustc_interface[a80cab810099d0cd]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>, rustc_driver[a177c53ac2cf0051]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[c08561276749af2a]::ErrorGuaranteed>>::{closure#1} as core[5571ea7c97e175b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  59:     0x7fb4bd02d5a8 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6bef02aba93c72af
  60:     0x7fb4bd068cb7 - std::sys::unix::thread::Thread::new::thread_start::hc79df79faa940a35
  61:     0x7fb4b7baffa3 - start_thread
                               at /build/glibc-fWwxX8/glibc-2.28/nptl/pthread_create.c:486:8
  62:     0x7fb4bceedeff - clone
                               at /build/glibc-fWwxX8/glibc-2.28/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
  63:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.64.0-dev running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [late_bound_vars_map] looking up late bound vars
#1 [explicit_predicates_of] computing explicit predicates of `_call`
#2 [collect_mod_item_types] collecting item types in top-level module
#3 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `research` due to previous error
```

File: src/test/incremental/hashes/inherent_impls.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pnkfelix) [reply src/test/incremental/hashes/inherent_impls.rs:122]: That is very subtle! I don't see any mention of it in the source for inherent_impls.rs itself.

I do see that pattern in other incremental tests, though for the most part it looked like the common pattern was to ensure that every comment line was the *same* length, while here for some reason this file is deliberately varying the line length...

Is there some place we can document this? Should we be trying to lint in some fashion to ensure this property is preserved going forward?

**Pull request merged**